### PR TITLE
[scala] Add support for CPD-ON and CPD-OFF special comments

### LINKED ISSUE #2480: [scala] Support CPD suppressions
As Scala is using Scala meta for parsing, and not a JavaCC / Antlr Grammar, it doesn't work on top of the base tokenizers / token filters, which means base functionality such as support for `CPD-OFF / CPD-ON` has to be built specially for this tokenizer, which is currently missing.

### PR DESCRIPTION
**Pull request by andyrobinson**
## Describe the PR

This enhancement adds support for CPD-ON and CPD-OFF special comments for Scala.  It does this by wrapping the scala.meta.tokens.Token with an adapter which satisfies the GenericToken interface, then uses the special comments processing from the existing BaseTokenFilter.

A new test fixture and results file has been added to the test, which seems consistent with the level of other testing in this area.

## Related issues

- Fixes #2480

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by travis)
- [x] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">No java rules are changed!</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-scala-modules/pmd-scala-common/src/main/java/net/sourceforge/pmd/cpd/ScalaTokenizer.java [hunk 1]
```diff
@@ -137,5 +155,27 @@ private boolean skipToken(Token token) {
             }
             return skip;
         }
+
+        private boolean isComment(Token token) {
+            return token != null && (token.text().startsWith("//") || token.text().startsWith("/*"));
```
Comment (oowekyala) [inline pmd-scala-modules/pmd-scala-common/src/main/java/net/sourceforge/pmd/cpd/ScalaTokenizer.java:160]:
```suggestion
            return token instanceof Token.Comment;
```

File: pmd-scala-modules/pmd-scala-common/src/test/resources/net/sourceforge/pmd/lang/scala/cpd/testdata/special_comments.txt [hunk 2]
```diff
@@ -0,0 +1,52 @@
+    [Image] or [Truncated image[            Bcol      Ecol
+L1
+    [// Testing CPD suppression]            1         27
+L3
+    [// Irrelevant comment]                 1         22
+L4
+    [// CPD-OFF]                            1         11
```
Comment (oowekyala) [inline pmd-scala-modules/pmd-scala-common/src/test/resources/net/sourceforge/pmd/lang/scala/cpd/testdata/special_comments.txt:7]: Shouldn't `CPD-OFF` tokens be filtered out too? In other languages I think that's how it works

File: pmd-scala-modules/pmd-scala-common/src/main/java/net/sourceforge/pmd/cpd/ScalaTokenAdapter.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    ScalaTokenAdapter(Token token, GenericToken comment) {
+        this.token = token;
+        this.previousComment = comment;
+    }
+
+    @Override
+    public GenericToken getNext() {
+        // not required
+        return null;
```
Comment (oowekyala) [inline pmd-scala-modules/pmd-scala-common/src/main/java/net/sourceforge/pmd/cpd/ScalaTokenAdapter.java:27]: Please either implement this or throw `NotImplementedException` (from lang3) or `UnsupportedOperationException`

File: pmd-scala-modules/pmd-scala-common/src/main/java/net/sourceforge/pmd/cpd/ScalaTokenizer.java [hunk 4]
```diff
@@ -125,7 +137,13 @@ Token getNextToken() {
...
             } while (token != null && skipToken(token) && tokenIter.hasNext());
 
-            return token;
+            result = new ScalaTokenAdapter(token, previousComment);
+
+            if (isComment(token)) {
+                previousComment = result;
+            }
+
+            return result;
```
Comment (oowekyala) [inline pmd-scala-modules/pmd-scala-common/src/main/java/net/sourceforge/pmd/cpd/ScalaTokenizer.java:146]: If the token is a comment it should actually not be returned, to implement this interface correctly. This is why your CPD-OFF comments are not filtered out (arguably all comments should actually be filtered out).

File: pmd-scala-modules/pmd-scala-common/src/main/java/net/sourceforge/pmd/cpd/ScalaTokenAdapter.java (hunk 3 above)
Comment (oowekyala) [reply pmd-scala-modules/pmd-scala-common/src/main/java/net/sourceforge/pmd/cpd/ScalaTokenAdapter.java:27]: Would also be nice to add a toString to this class (any autogenerated one will do)

**Review by oowekyala — COMMENTED** [review]
Hi, thanks for working on this! I think it's ready enough to schedule for the next release (around next weekend). If you can't do those changes by then I'll do them when merging

File: pmd-scala-modules/pmd-scala-common/src/main/java/net/sourceforge/pmd/cpd/ScalaTokenizer.java (hunk 1 above)
Comment (andyrobinson) [reply pmd-scala-modules/pmd-scala-common/src/main/java/net/sourceforge/pmd/cpd/ScalaTokenizer.java]: Done

File: pmd-scala-modules/pmd-scala-common/src/test/resources/net/sourceforge/pmd/lang/scala/cpd/testdata/special_comments.txt (hunk 2 above)
Comment (andyrobinson) [reply pmd-scala-modules/pmd-scala-common/src/test/resources/net/sourceforge/pmd/lang/scala/cpd/testdata/special_comments.txt]: Comments are now skipped and no longer tokenised

File: pmd-scala-modules/pmd-scala-common/src/main/java/net/sourceforge/pmd/cpd/ScalaTokenAdapter.java (hunk 3 above)
Comment (andyrobinson) [reply pmd-scala-modules/pmd-scala-common/src/main/java/net/sourceforge/pmd/cpd/ScalaTokenAdapter.java]: now throws

File: pmd-scala-modules/pmd-scala-common/src/main/java/net/sourceforge/pmd/cpd/ScalaTokenizer.java (hunk 4 above)
Comment (andyrobinson) [reply pmd-scala-modules/pmd-scala-common/src/main/java/net/sourceforge/pmd/cpd/ScalaTokenizer.java]: rewritten to skip comments

**Comment by andyrobinson** [comment]
Ok, done the changes, and comments are no longer tokenised (this is a change from the original tokenisation implementation).    Thanks for the comments.

File: pmd-scala-modules/pmd-scala-common/src/main/java/net/sourceforge/pmd/cpd/ScalaTokenizer.java [hunk 5]
```diff
@@ -103,29 +108,39 @@ public void tokenize(SourceCode sourceCode, Tokens tokenEntries) throws IOExcept
...
 
     /**
-     * Token Filter skips un-helpful tokens to only register important tokens
+     * Implementation of the generic Token Manager, also skips un-helpful tokens and comments to only register important tokens
      * and patterns.
+     *
+     * Keeps track of comments, for special comment processing
      */
-    private static class ScalaTokenFilter {
+    private class ScalaTokenManager implements TokenManager {
```
Comment (adangel) [inline pmd-scala-modules/pmd-scala-common/src/main/java/net/sourceforge/pmd/cpd/ScalaTokenizer.java:116]: I think, this class can be static

File: pmd-scala-modules/pmd-scala-common/src/main/java/net/sourceforge/pmd/cpd/ScalaTokenizer.java [hunk 6]
```diff
@@ -137,5 +152,27 @@ private boolean skipToken(Token token) {
...
+            return token instanceof Token.Comment;
+        }
+
+        @Override
+        public void setFileName(String fileName) {
+            throw new UnsupportedOperationException("setFileName deprecated");
+        }
+    }
+
+    private class ScalaTokenFilter extends BaseTokenFilter<ScalaTokenAdapter> {
```
Comment (adangel) [inline pmd-scala-modules/pmd-scala-common/src/main/java/net/sourceforge/pmd/cpd/ScalaTokenizer.java:166]: can be static as well

**Review by adangel — APPROVED** [review]
Thanks!

I'll make the inner classes static and merge for 6.30.0

**Pull request merged**